feat: add base inputs, types and pydantic models for validate image project type

tnagorra opened the pull request:
## Depends on

- https://github.com/mapswipe/mapswipe-firebase/pull/3
- https://github.com/mapswipe/mapswipe-firebase/pull/6

## Changes

- Add base inputs, types and pydantic models for validate image type
- Update todo
- Update firebase to latest develop

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [ ] tests
- [x] permission checks
- [x] translations

tnagorra committed:
feat(validate-image): add base inputs, types and pydantic models

- Update todo
- Update firebase to latest develop

tnagorra committed:
feat(firebase): add handlers for creating and updating a project in firebase

- add fields on project to track firebase sync
- add pyfirebase_mapswipe library import
- add task to push project data to firebase
- update processed project serializer to trigger task

tnagorra commented on Dockerfile:
```diff
@@ -10,8 +10,13 @@ ENV UV_COMPILE_BYTECODE=1
 ENV UV_PROJECT_ENVIRONMENT="/usr/local/"
 
+COPY firebase /code/firebase
+
```
Let's remove this.

tnagorra commented on project_types/base/project.py:
```diff
@@ -31,6 +37,8 @@ def get_max_time_spend_percentile(project_type: ProjectTypeEnum) -> float:
             return 11.2
         case ProjectTypeEnum.VALIDATE:
             return 6.1
+        case ProjectTypeEnum.VALIDATE_IMAGE:
```
Let's add a TODO

tnagorra commented on project_types/base/project.py:
```diff
@@ -161,3 +171,55 @@ def process_project(self) -> bool:
+                self.handle_project_update_on_firebase(project_ref, valid_project)
+        except Exception as ex:
+            self.project.update_firebase_push_status(FirebasePushStatusEnum.FAILED)
+            raise ex from ex
```
- Catch specific exceptions.
- Use logger for rest of the cases

tnagorra commented on project_types/store.py:
```diff
@@ -36,12 +41,15 @@ def get_project_property(project_type: ProjectTypeEnum | None):
     if project_type == ProjectTypeEnum.VALIDATE:
         return ("validate", ValidateProjectProperty)
+    if project_type == ProjectTypeEnum.VALIDATE_IMAGE:
+        # TODO(tnagorra): Need to confirm if validate_image or validateImage
```
Let's remove this comment

tnagorra commented on project_types/store.py:
```diff
@@ -22,6 +24,9 @@ def get_tutorial_task_property(project_type: ProjectTypeEnum | None):
     if project_type == ProjectTypeEnum.VALIDATE:
         return ("validate", validate_tutorial.ValidateTutorialTaskProperty)
+    if project_type == ProjectTypeEnum.VALIDATE_IMAGE:
+        # TODO(tnagorra): Need to confirm if validate_image or validateImage
```
Let's remove this comment

tnagorra commented on project_types/tile_map_service/find/project.py:
```diff
@@ -1,7 +1,63 @@
+    return obj
+
+
+# TODO(tnagorra): Move this somewhere more appropriate
+def project_type_enum_to_firebase(input_enum: ProjectTypeEnum) -> firebase_models.FbEnumProjectType:
+    match input_enum:
+        case ProjectTypeEnum.FIND:
+            return firebase_models.FbEnumProjectType.FIND
+        case ProjectTypeEnum.COMPARE:
+            return firebase_models.FbEnumProjectType.COMPARE
+        case ProjectTypeEnum.COMPLETENESS:
+            return firebase_models.FbEnumProjectType.COMPLETENESS
+        case ProjectTypeEnum.VALIDATE:
+            return firebase_models.FbEnumProjectType.VALIDATE
+        case ProjectTypeEnum.VALIDATE_IMAGE:
+            return firebase_models.FbEnumProjectType.VALIDATE_IMAGE
+
+
+# TODO(tnagorra): Move this somewhere more appropriate
+def raster_tile_server_name_enum_to_firebase(
+    input_enum: RasterTileServerNameEnum,
+) -> firebase_models.FbEnumRasterTileServerName:
+    match input_enum:
+        case RasterTileServerNameEnum.CUSTOM:
+            return firebase_models.FbEnumRasterTileServerName.CUSTOM
+        case RasterTileServerNameEnum.BING:
+            return firebase_models.FbEnumRasterTileServerName.BING
+        case RasterTileServerNameEnum.MAPBOX:
+            return firebase_models.FbEnumRasterTileServerName.MAPBOX
+        case RasterTileServerNameEnum.MAXAR_STANDARD:
+            return firebase_models.FbEnumRasterTileServerName.MAXAR_STANDARD
+        case RasterTileServerNameEnum.MAXAR_PREMIUM:
+            return firebase_models.FbEnumRasterTileServerName.MAXAR_PREMIUM
+        case RasterTileServerNameEnum.ESRI:
+            return firebase_models.FbEnumRasterTileServerName.ESRI
+        case RasterTileServerNameEnum.ESRI_BETA:
+            return firebase_models.FbEnumRasterTileServerName.ESRI_BETA
```
Move this to project_types/firebase.py

tnagorra commented on project_types/tile_map_service/find/project.py:
```diff
@@ -28,3 +84,121 @@ def __init__(self, project: Project):
+
+    @typing.override
+    def handle_new_project_on_firebase(self, project_ref):
+        assert self.project.tutorial_id is not None
+        assert self.project.image is not None
```
Let's add messages.

tnagorra committed:
fix(models): add old_id on tutorial and user model

tnagorra committed:
refactor(firebase): move core functionality to base project

- define get_project_specifics_for_firebase for find project
- use old_id instead of id when syncing models with firebase
- handle specific exception when handling invalid project config
- move common functions to project_types/firebase

thenav56 commented on apps/user/models.py:
```diff
@@ -14,6 +14,8 @@ class User(AbstractUser):
     email = models.EmailField(unique=True)
     display_name = models.CharField(max_length=255)
+    # FIXME(tnagorra): We might need to skip the indexing
+    old_id = models.CharField(max_length=30, db_index=True, null=True)
```
Let's rename this to firebase_userid

tnagorra replied:
That makes sense!

tnagorra replied:
I have added a FIXME for the time being. Will add this once all of the other PRs are merged :)

thenav56 commented on apps/project/tasks.py:
```diff
@@ -25,3 +25,20 @@ def process_project_task(project_id: int):
+            return None
+
+    project = Project.objects.get(pk=project_id)
+    try:
+        project_type_handler = get_project_type_handler(project.project_type_enum)(project)
+        project_type_handler.push_to_firebase()
+        return True
+    except Exception:
+        logger.error("Push project to firebase failed", exc_info=True)
+        return False
```
We can now remove the try/except from here as it is already handled by push_to_firebase

tnagorra replied:
Done!

thenav56 commented on project_types/tile_map_service/compare/project.py:
```diff
@@ -74,3 +76,9 @@ def create_tasks(self, group: ProjectTaskGroup, raw_group: tile_grouping.RawGrou
+
+    @typing.override
+    def get_project_specifics_for_firebase(self, project_ref):
+        class EmptyModel(BaseModel): ...
+
+        return EmptyModel()
```
Maybe we should add ProjectSpecificFirebaseEmptyModel in the base (or in firebase models) and use that?
It will be easy to check which projects are using this.

tnagorra replied:
This will be very temporary, so we might not need to have a common empty class.

I have already replaced most of these usecases on another branch.

tnagorra replied:
Done!

thenav56 reviewed (dismissed):
LGTM

susilnem commented on apps/project/tasks.py:
```diff
@@ -25,3 +25,20 @@ def process_project_task(project_id: int):
+            return None
+
+    project = Project.objects.get(pk=project_id)
+    try:
```
We don't need this here.

susilnem replied:
Already mentioned here! https://github.com/mapswipe/mapswipe-backend/pull/50#discussion_r2161042765
So resolving this comment!

susilnem commented on project_types/validate_image/tutorial.py:
```diff
@@ -0,0 +1,10 @@
+from project_types.base.tutorial import BaseTutorialTaskProperty
+from utils import fields as custom_fields
+
+logger = logging.getLogger(__name__)
```
```suggestion

```

tnagorra replied:
Any reason to remove logger?

thenav56 replied:
It's fine for `logger`

susilnem commented on project_types/validate_image/tutorial.py:
```diff
@@ -0,0 +1,10 @@
+import logging
```
```suggestion
```

tnagorra replied:
It's fine for `logger`

tnagorra committed:
feat(firebase): add project specific handlers for compare and completeness

tnagorra committed:
feat(custom_options): add custom options to validate and validate_image

tnagorra committed:
feat(firebase): setup test environment for firebase

- Add service test and firebase-test
- Clear firebase data if running on emulator and whitelisted

tnagorra committed:
feat(firebase): sync tasks and groups on firebase

- add fallback empty models for task, group and project specifics
- calculate required_count for group
- calculate required_results for project

tnagorra committed:
feat(firebase): implement project specifics for project, task and group

- add project creation on firebase: validate, validate_image
- add task/groups creation on firebase: find, compare, completeness, validate, validate_image

tnagorra committed:
feat(ci): delegate builds to bake for better perf

thenav56 committed:
fix: update workflow to use new test container

- add firebase build using docker build action
